src/main/java/org/breedinginsight/brapps/importer/controllers/UploadController.java

```diff
@@ -0,0 +1,157 @@
+import org.breedinginsight.api.auth.SecurityService;
+import org.breedinginsight.api.model.v1.response.Response;
+import org.breedinginsight.api.v1.controller.metadata.AddMetadata;
+import org.breedinginsight.brapps.importer.model.response.ImportResponse;
+import org.breedinginsight.brapps.importer.services.FileImportService;
+import org.breedinginsight.services.exceptions.AuthorizationException;
+import org.breedinginsight.services.exceptions.DoesNotExistException;
+import org.breedinginsight.services.exceptions.UnprocessableEntityException;
+import org.breedinginsight.services.exceptions.UnsupportedTypeException;
+
+import javax.inject.Inject;
+import java.util.UUID;
+
+@Slf4j
+@Controller("/${micronaut.bi.api.version}")
+public class UploadController {
+    private SecurityService securityService;
+    private FileImportService fileImportService;
+
+    @Inject
+    UploadController(SecurityService securityService, FileImportService fileImportService) {
+        this.securityService = securityService;
+        this.fileImportService = fileImportService;
+    }
+
+    @Post("programs/{programId}/import/mappings/{mappingId}/data")
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    @Produces(MediaType.APPLICATION_JSON)
+    @AddMetadata
+    @ProgramSecured(roles = {ProgramSecuredRole.BREEDER, ProgramSecuredRole.SYSTEM_ADMIN})
+    public HttpResponse<Response<ImportResponse>> uploadData(@PathVariable UUID programId, @PathVariable UUID mappingId,
+                                                             @Part("file") CompletedFileUpload file) {
+        try {
+            AuthenticatedUser actingUser = securityService.getUser();
+            ImportResponse result = fileImportService.uploadData(programId, mappingId, actingUser, file);
+            Response<ImportResponse> response = new Response(result);
+            return HttpResponse.ok(response);
+        } catch (DoesNotExistException e) {
+            log.info(e.getMessage());
+            return HttpResponse.notFound();
+        } catch (AuthorizationException e) {
+            log.info(e.getMessage());
+            return HttpResponse.status(HttpStatus.FORBIDDEN, e.getMessage());
+        } catch (UnsupportedTypeException e) {
+            log.info(e.getMessage());
+            return HttpResponse.status(HttpStatus.UNSUPPORTED_MEDIA_TYPE, e.getMessage());
+        } catch (UnprocessableEntityException e) {
+            log.info(e.getMessage());
+            return HttpResponse.status(HttpStatus.UNPROCESSABLE_ENTITY, e.getMessage());
+        }
+    }
+
+
+    @Get("programs/{programId}/import/mappings/{mappingId}/data/{uploadId}{?mapping}")
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    @Produces(MediaType.APPLICATION_JSON)
+    @AddMetadata
+    @ProgramSecured(roles = {ProgramSecuredRole.BREEDER, ProgramSecuredRole.SYSTEM_ADMIN})
+    public HttpResponse<Response<ImportResponse>> uploadData(@PathVariable UUID programId, @PathVariable UUID mappingId,
```

Comment:
Could this method name be changed? Maybe to `getUploadData`

Reply:
Changed and pushed